include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,436 @@
+ *   to supply a desired allocator. Use Make*() format specific interfaces to supply values and format
+ *   specific indices. The specified data will be copied into an internally allocated buffer.
+     Internally, we will represent a SparseTensor as a single contiguous buffer that
+ *   contains values followed by format specific indices. We use Tensors to project
+ *   values and indices into various parts of buffer.
  */
 
 class SparseTensor final {
  public:
+  /// <summary>
+  /// This constructs an instance that points to user defined buffers.
+  /// Make use of Use* functions to supply format specific indices that
+  /// reside in the user supplied buffers. The instance constructed this way
+  /// will not copy data and will use the supplied buffers in read-only way.
+  /// The lifespan of supplied buffers is expected to eclipse the lifespan of the
+  /// sparse tensor instance.
+  /// </summary>
+  /// <param name="elt_type">MlDataType</param>
+  /// <param name="dense_shape">a shape of original tensor in dense form</param>
+  /// <param name="values_shape">shape for user supplied values</param>
+  /// <param name="values_data">a pointer to values</param>
+  /// <param name="location">description of the user allocated memory</param>
   SparseTensor(MLDataType elt_type,
-               const TensorShape& shape,
-               size_t nnz,
+               const TensorShape& dense_shape,
+               const TensorShape& values_shape,
                void* values_data,
-               void* indices_data,
-               const OrtMemoryInfo& memory_info);
+               const OrtMemoryInfo& location);
 
+  /// <summary>
+  /// Use this constructor to hold sparse data in the buffer
+  /// allocated with the specificed allocator. Use Make*() methods
+  /// to populate the instance with data which will be copied into the
+  /// allocated buffer.
+  /// </summary>
+  /// <param name="elt_type"></param>
+  /// <param name="dense_shape"></param>
+  /// <param name="allocator"></param>
   SparseTensor(MLDataType elt_type,
-               const TensorShape& shape,
-               size_t nnz,
+               const TensorShape& dense_shape,
                std::shared_ptr<IAllocator> allocator);
 
-  ~SparseTensor() = default;
+  SparseTensor() noexcept;
+
+  ~SparseTensor();
 
-  // For now, disallow all copy, assignment, and move.
-  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(SparseTensor);
+  // For now, disallow copy and assignment.
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(SparseTensor);
 
-  // Returns the number of entries in the values tensor (aka "NNZ" or "number of nonzero values")
+  /// <summary>
+  // Returns the number of non-zero values (aka "NNZ")
+  // For block sparse formats this may include some zeros in the blocks
+  // are considered non-zero.
+  /// </summary>
+  /// <returns>nnz</returns>
   size_t NumValues() const { return static_cast<size_t>(values_.Shape().Size()); }
 
```

Comment:
Why do we convert the int64_t from Shape().Size() to size_t here?

Reply:
The choice was made in favor of least casts. The original interface also returned size_t and that seems more convinient in many circumstances.